### DISCUSSION TITLE
IIoT Edge Devices --> arm based Systems

### DISCUSSION THREAD

**Discussion by akrea** [eclipse-kura/kura#4832] [General]
Hi

First of all: THANK YOU providing this tool and the effort the contributors put into this.

When I checked on the Kura/Kapua combo, the most glaring missing piece I see is that there is no arm-based container version of Kura. In the industrial manufacturing environment we have a vast majority of edge devices which are arm based and run services as docker containers (currently moving from RPi4 (32bit) to i.MX 8M  (64bit)).

My question here is: is there no arm-container-version due to resource constraints in the dev-team or do I miss an (obvious) reason why there is no arm-container-version?

I not familiar in coding for arm architecture nor do I have a lot of experience in creating Docker images. Though I'm willing to learn, try and if I succeed to share :-) If you could provide some starting guidance (and a rough estimate on how much time it would take you to do it - just to get an idea what I'm getting into) I would be delighted. PM if you prefer.

Again: Thank you!

**Comment by mattdibi** [comment]
Hello there @akrea, thanks for reaching out.

> is there no arm-container-version due to resource constraints in the dev-team or do I miss an (obvious) reason why there is no arm-container-version?

It's mainly a resource constraint issue. We already have all the pieces (i.e. we already support Kura on ARM devices natively) we just need to package them into a container.

The other constraint I see is related to the Jenkins pipeline responsible for building Kura: it's x86-64 only and we would need some changes to support ARM Docker images builds 🤔 

> If you could provide some starting guidance (and a rough estimate on how much time it would take you to do it - just to get an idea what I'm getting into) I would be delighted. PM if you prefer.

I think @GregoryIvo can help you with this. He's responsible for the latest changes in that department and can provide some useful info.

**Comment by GregoryIvo** [comment]
I think it would be really neat to have container support for different architectures, and think this would be very achievable 😄 

I personally would love to see this support! 🚀 🌕 

 ---

## How our build system for containers works:

we have a master POM file, that builds all of our Kura profiles. To build kura containers the following occurs: 

1) POM first builds a kura 'base' image, which is used as a resource for other containers. (This basically just compiles kura for the architecture)

https://github.com/eclipse/kura/blob/60fa70ce72434274db9cab391fdd0ec13fb1732b/kura/distrib/pom.xml#L2048C41-L2066

2) We then Take the base image and use it as a resource to build specific images like UBI8  (basically we take the compiled 'binary' from before and install it in a container with an os) 

https://github.com/eclipse/kura/blob/60fa70ce72434274db9cab391fdd0ec13fb1732b/kura/distrib/pom.xml#L2102-L2142

3) now you are wondering where the Dockerfile is for these containers. They can be found here:
Base image: https://github.com/eclipse/kura/tree/develop/kura/distrib/src/main/resources/docker-x86_64-nn
Ubi-8: https://github.com/eclipse/kura/tree/develop/kura/distrib/src/main/resources/docker-ubi8-x86_64-nn

---

## What changes would need to be made

In my opinion, this project is really doable. 

1) make a arm32/aarch64 Profile using the respective generic profile.  (Basically, duplicate the x86 one, and change x86 to the desired arch)

2) add entry to the pom for the new arch base image**

3)Duplicate the ubi8 and Alpine x86 container profiles, and change them to the desired arch 

4) add entry to the pom for these new containers**

**NOTE: our Jenkins pipeline runs on x86, so we must use something like BuildX in docker to cross-compile. 

Bonus points: Currently all containers are built with the flag buildAllContainers, it would be nice to keep this flag, and also add buildAllContainers<Arch>

**Comment by akrea** [comment]
Thank you for the guidance. I'm currently setting up a Sparkplug 3 testing environment in Node-Red (as I need to educate myself on this too) 🤓. As soon as this is done (by mid September at the latest) I will takle this challenge. May I come back for advice if needed?

**Reply by mattdibi** [reply]
> May I come back for advice if needed?

Yes, of course. Feel free to ping us here in this discussion :)

**Reply by akrea** [reply]
It'll take a little longer to get going as life threw me some minor curve balls...

**Comment by akrea** [comment]
In order to evaluate if the image is working, I fired up an existing docker image on a ubuntu server to see how kura is supposed to work and familiarize myself with it.

When I open the documentation I see that there should be a network and a firewall section within the "system"-part. I can imagine that when using docker those two are managed by docker and do not appear (same goes for container orchestration in "service"-part). But then i tried to setup the heater demo. While I was able to install the package successfully (according to logs), the heater does not show up in the Services. Moreover, when I try to install a driver, the Driver Factory --- Select Component Factory --- list is empty. The only warnings I see in the logs are about skins...

On the positive side - I was able to fire up Kapua, connect Kura to the Broker and the connection shows up in Kapua.

Am I missing something or is Kura it supposed to work this way?

**Comment by mattdibi** [comment]
> When I open the documentation I see that there should be a network and a firewall section within the "system"-part. I can imagine that when using docker those two are managed by docker and do not appear (same goes for container orchestration in "service"-part).

That's correct. The dockerized version of Kura doesn't manage the network yet.

> But then i tried to setup the heater demo. While I was able to install the package successfully (according to logs), the heater does not show up in the Services

This is not expected. Can you provide a full log? Did you check for additional info in the `kura-console.log` file under `/var/logs`?

> Moreover, when I try to install a driver, the Driver Factory --- Select Component Factory --- list is empty. 

Again. Doesn't sound right. Check again in the logs and the `kura-console.log` for additional errors.

> On the positive side - I was able to fire up Kapua, connect Kura to the Broker and the connection shows up in Kapua.

This is great news!

**Reply by GregoryIvo** [reply]
Amazing news @akrea! I'm excited to see the PR

**Comment by akrea** [comment]
Sorry for the wait - was on vacation and needed family time.

AND the time off was good to have a fresh mind:

I did a clean restart of Kura. For the heater I chose the wrong version the first time and didn't notice it and for the driver: i figured that the have to be separately loaded before they appear in the list. Typical case of [PEBCAK](https://en.wiktionary.org/wiki/PEBCAK).

Going forward with some testing.

**Reply by akrea** [reply]
I have now been trying to get data published to kapua by the heater demo and by the example publisher (for two hours). To no avail. There are simply no data arriving in kapua (I also cannot see anything on the broker with mqtt explorer). Although I do see the birth certificate of the device in kapua. Can it be that hard to get it talking?

I do not have any log entries from the example publisher.  The heater tells me, that I do not have a "publisher" selected, so it cannot publish. This is simply wrong as I have set up and selected one and applied the changes (several times).

Any ideas? I would like to get it working before jumping into building images as I cannot judge if a image is flawed if I don't know how to get it working in the first place.

**Reply by MMaiero** [reply]
Hi,
can you please post the configuration of Kura you are using?
Can you also post the logs on the container? Just to understand where the problem is and where to start digging with the investigation

**Reply by akrea** [reply]
Attached the Logs and the snapshot. Hope this helps.

[Kura_Logs (1).zip](https://github.com/eclipse/kura/files/12907989/Kura_Logs.1.zip)
[snapshot_1697315769363.json](https://github.com/eclipse/kura/files/12907991/snapshot_1697315769363.json)
[container.log](https://github.com/eclipse/kura/files/12907997/container.log)

**Reply by MMaiero** [reply]
Thanks.
The publishing seems to be working. Just for the sake of testing, did you also try to configure the publisher with QoS 1 on Kura side?

Also, can you share the Kapua setup and configuration you are using? What accounts, what is their configuration, etc. So that we can recreate the environment.

**Reply by akrea** [reply]
Sorry it always takes so long to answer. There is just too much going on!

No, changing to QoS 1 did not change the outcome. I can see that the topic tree is created (but that is true for QoS0 too) but no data is fed into it.

In Kapua I did not see any convenient way to get a "snapshot" of the current config. But there is not much my installation. I did docker deploy according to [the readme](https://github.com/eclipse/kapua/blob/develop/README.md) on a vanilla Ubuntu 22.04 Server. The I played around a little, thats when I posted on September, 26th. After that I followed [this guide](https://github.com/eclipse/kapua/blob/develop/docs/kuraKapuaDocs.md) to get the demo working. Actually I followed a youtube video implementing this guide.

If you can point me to the files I can provide you in order to get the info you need - let me know.

Kapua Build is 1.6.9-45cb123fccbf39d8539a01cbad241fd39459051f

**Reply by MMaiero** [reply]
Thanks.
I have asked you about the qos 1 to try to determine where the issue is.
If you can post the Kura logs, we can see if the message is confirmed. If so, we are sure that the message is delivered and we can look at understanding if there are issues with the Kapua setup.
If not, we will concentrate on the Kura side.

**Reply by akrea** [reply]
Sure, here you go!

[Kura_Logs.zip](https://github.com/eclipse/kura/files/13254866/Kura_Logs.zip)

**Reply by MMaiero** [reply]
Ok, so from the logs it seems that the messages are properly sent from Kura and confirmed by Kapua as you can see here:

```
2023-11-03T18:54:21,419 [DataServiceImpl:Submit] INFO  o.e.k.c.d.t.m.MqttDataTransport - Publishing message on topic: Account123/kura1ID/W1/A1/testasset with QoS: 1
2023-11-03T18:54:21,423 [MQTT Call: kura1ID] INFO  o.e.k.c.d.DataServiceImpl - Confirmed message ID: 255627 to store
2023-11-03T18:54:21,424 [pool-24-thread-1] INFO  o.e.k.e.p.ExamplePublisher - Confirmed message with id: 255627
```

So, maybe the problem is in the Kapua side.
Can you please post the configuration you have there?
Are you publishing in a root account or in a child account?

**Reply by akrea** [reply]
As I said

- I just cloned the kapua repo and started the script. I had to change the `docker-compse` commands to `docker compose` as I use a newer version of docker. But I did not change anything else in the config. I will galdly post my config but I did not find a way to export configs. Let me know what exactly you need. I did screenshots the about info - if this helps...
- I followed [this guide](https://github.com/eclipse/kapua/blob/develop/docs/kuraKapuaDocs.md) to set up Account123 and User123 (see last screenshot).

![kapua1](https://github.com/eclipse/kura/assets/16197061/61c8c41a-03c6-4181-9a1d-70906a0db987)
![kapua2](https://github.com/eclipse/kura/assets/16197061/7fbe3e97-e425-48e5-8f86-d23dbda0eda7)
![kapua3](https://github.com/eclipse/kura/assets/16197061/8a3600a1-056b-4c7c-8ed7-d13760049cfa)
![kapua4](https://github.com/eclipse/kura/assets/16197061/a9861c7a-1fff-4170-9817-04fe84b875ec)
![kapua5](https://github.com/eclipse/kura/assets/16197061/fa40ea85-ed69-4c0f-aad9-f207c90acf00)
![kapua6](https://github.com/eclipse/kura/assets/16197061/899f80a5-d87f-45c1-bb43-b3c99f54d86c)
![kapua7](https://github.com/eclipse/kura/assets/16197061/d758a5a0-3968-4c97-83e1-162b66bb2240)
![kapua8](https://github.com/eclipse/kura/assets/16197061/8a8c0a0b-06c6-49d4-a4c3-11f81d0a703d)
![kapua9](https://github.com/eclipse/kura/assets/16197061/9fddd699-ec55-4ee3-a2c4-2cce3211e76f)
![Screenshot 2023-11-05 203443](https://github.com/eclipse/kura/assets/16197061/f7826ccb-9f1c-4802-8c0d-e80df3f70610)

**Comment by Coduz** [comment]
Hi @akrea ,

I'm from the [eclipse/kapua](https://github.com/eclipse/kapua) project.

May I ask you which version of Kapua are you running? 
You can find it in the bottom right corner of the Web UI. 

Regards, 

Alberto

**Reply by akrea** [reply]
Hi @Coduz 

Thanks for your efforts.

> Kapua Build is 1.6.9-45cb123fccbf39d8539a01cbad241fd39459051f

**Reply by Coduz** [reply]
Do you have any access to logs? Especially the one from the `broker` and `es`  container?
